ocfweb/docs/views/servers.py

```diff
@@ -57,81 +56,90 @@ def has_munin(self):
-                    'tsunami',
-                    'werewolves',
-                    'whirlwind',
-                    'whiteout',
-                    'zombies',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='pandemic',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'pileup',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='riptide',
-            type='hypervisor',
-        ),
-
-        Host.from_ldap(
-            hostname='corruption',
-            type='server',
-        ),
-
-        Host('blackhole', 'network', 'Managed Cisco Catalyst 2960S-48TS-L Switch.', []),
-
-        Host('logjam', 'printer', '', []),
-        Host('pagefault', 'printer', '', []),
-        Host('papercut', 'printer', '', []),
-        Host.from_ldap('overheat', type='raspi'),
-        Host.from_ldap('tornado', type='nuc'),
-
-        *(Host.from_ldap(desktop, type='desktop')
-          for desktop in sorted(list_desktops())),
-    ]
+    def create_hosts(lst):
+        """Accepts a list of raw ldap output, returns a dictionary of host
+        objects indexed by hostname
+        """
+        hosts = {}
+        for h in lst:
+            if not is_hidden(h):
+                description = h.get('description', [''])[0]
+                hostname = h['cn'][0]
+                hosts[hostname] = (Host(hostname, h['type'], description, ()))
+        return hosts
+
+    servers = create_hosts(hosts_by_filter('(type=server)'))
+    desktops = create_hosts(hosts_by_filter('(type=desktop)'))
+    misc = create_hosts(hosts_by_filter('(type=printer)'))
+    hypervisors = {}
+
+    physical_hosts = query_puppet(TEMP_IS_VIRT_FALSE)
+    all_children = query_puppet(TEMP_GET_CHILDREN)
+
+    # Add children to hypervisors
+    for h in list(servers.values()):
+        if h.hostname in physical_hosts:
+            # Populate a list of children
+            children = []
+            for child_hostname in all_children.get(h.hostname, []):
+                child = servers.get(child_hostname, False)
+                if child:
+                    del servers[child.hostname]
+                    children.append(Host(child.hostname, 'vm', child.description, ()))
+            # Associate host with its children and move it to the hypervisors dictionary
+            del servers[h.hostname]
+            hypervisors[h.hostname] = Host(
+                h.hostname,
+                'hypervisor',
+                h.description,
+                children,
+            )
+
+    # Handle special cases
+    def change_host_type(hostname, host_type, origin_type_dict, target_type_dict):
+        """Pop Host with hostname from servers, change its type, and add to misc
+        """
+        del origin_type_dict[hostname]
+        target_type_dict[hostname] = Host.from_ldap(hostname, type=host_type)
```

Comment:
I would suggest the following pattern:

```python
host = origin_type_dict.pop(hostname)
target_type_dict[hostname] = host._replace(type=host_type)
```

It would avoid making an extra LDAP query.